prov/efa/src/rdm/rxr_pkt_pool.c

```diff
@@ -115,24 +184,144 @@ int rxr_pkt_pool_grow(struct rxr_pkt_pool *rxr_pkt_pool)
-		return err;
+	if (rxr_pkt_pool->sendv_pool) {
+		err = ofi_bufpool_grow(rxr_pkt_pool->sendv_pool);
+		if (err)
+			return err;
+	}
 
 	return 0;
 }
 
+/**
+ * @brief post multiple packet entries to receive messages from shm
+ *
+ * @param[in]		pkt_pool	pkt pool, whose usage must be SHM RX packet pool
+ * @param[in]		nrecv		number of receiving buffers to post
+ * @returns
+ * On success, return 0.
+ * On failure, return negative libfabric error code.
+ */
+static inline
+int rxr_pkt_pool_bulk_post_recv_shm(struct rxr_pkt_pool *pkt_pool, int nrecv)
+{
+	int irecv, err;
+	struct rxr_pkt_entry *pkt_entry;
+
+	assert(pkt_pool->usage == RXR_PKT_FROM_SHM_RX_POOL);
+
+	for (irecv = 0; irecv < nrecv; ++irecv) {
+		pkt_entry = rxr_pkt_entry_alloc(pkt_pool->ep, pkt_pool, RXR_PKT_FROM_SHM_RX_POOL);
+		pkt_entry->x_entry = NULL;
+
+		err = fi_recv(pkt_pool->ep->shm_ep, pkt_entry->wiredata, pkt_pool->ep->mtu_size, NULL, FI_ADDR_UNSPEC, pkt_entry);
+		if (OFI_UNLIKELY(err)) {
+			rxr_pkt_entry_release_rx(pkt_pool->ep, pkt_entry);
+			EFA_WARN(FI_LOG_EP_CTRL,
+				"failed to post buf for shm  %d (%s)\n", -err,
+				fi_strerror(-err));
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief post multiple packet entries to receive messages from EFA device
+ *
+ * @param[in]		pkt_pool	pkt pool, whose usage must be EFA RX packet pool
+ * @param[in]		nrecv		number of receiving buffers to post
+ * @returns
+ * On success, return 0.
+ * On failure, return negative libfabric error code.
+ */
+static inline
+int rxr_pkt_pool_bulk_post_recv_efa(struct rxr_pkt_pool *pkt_pool, int nrecv)
+{
+	int irecv, err;
+	struct rxr_pkt_entry *pkt_entry;
+	struct ibv_recv_wr *prv_recv_wr, *cur_recv_wr, *bad_wr;
+
+	assert(pkt_pool->usage == RXR_PKT_FROM_EFA_RX_POOL);
+	assert(pkt_pool->efa_recv_wr_pool);
+
+	prv_recv_wr = NULL;
+	for (irecv = 0; irecv < nrecv; ++irecv) {
+		pkt_entry = rxr_pkt_entry_alloc(pkt_pool->ep, pkt_pool, RXR_PKT_FROM_EFA_RX_POOL);
+		if (pkt_entry == NULL) {
+			EFA_WARN(FI_LOG_CQ, "allocating EFA rx pakcet entry failed!");
+			return -FI_ENOBUFS;
+		}
+
+		pkt_entry->x_entry = NULL;
+
+		cur_recv_wr = &pkt_pool->efa_recv_wr_pool[irecv].wr;
+		cur_recv_wr->next = NULL;
+
+		cur_recv_wr->wr_id = (uintptr_t)pkt_entry;
+		cur_recv_wr->num_sge = 1;	/* Always post one iov/SGE for recv */
+
+		cur_recv_wr->sg_list = pkt_pool->efa_recv_wr_pool[irecv].sge;
+		cur_recv_wr->sg_list[0].length = pkt_pool->ep->mtu_size;
+		assert(pkt_entry->mr);
+		cur_recv_wr->sg_list[0].lkey = ((struct efa_mr *)pkt_entry->mr)->ibv_mr->lkey;
+		cur_recv_wr->sg_list[0].addr = (uintptr_t)pkt_entry->wiredata;
+
+		if (irecv > 0)
+			prv_recv_wr->next = cur_recv_wr;
+
+		prv_recv_wr = cur_recv_wr;
+	}
+
+#if HAVE_LTTNG
+	{
+		struct ibv_recv_wr *work = head_recv_wr;
+		while (work) {
+			efa_tracepoint_wr_id_post_recv((void *) work->wr_id);
+			work = work->next;
+		}
+	}
+#endif
+
+	err = ibv_post_recv(pkt_pool->ep->base_ep.qp->ibv_qp, &pkt_pool->efa_recv_wr_pool[0].wr, &bad_wr);
```

Comment:
Why is `&pkt_pool->efa_recv_wr_pool[0].wr` acceptable? It used to be `head_recv_wr`

Reply:
`head_recv_wr` is a pointer pointing to `&pkt_pool->efa_recv_wr_pool[0].wr`